### ISSUE TITLE
semantics of Species attributes and their use in (non)specific models

### ISSUE THREAD

**Issue by grahamgower** [popgensims/stdpopsim#243]
Currently, the `population_size_citations` and `generation_time_citations` attributes of a given species are not being written with the other citations. Ideally they should be, but this raises the question of when these citations are relevant to a given simulation.

Consider two homsap models. `ooa_3` uses the species defined generation_time of 25. Whereas `ooa_archaic` defines generation_time to be 29. In the former case, the species generation_time_citations are relevant, in the latter they are not (ignoring of course that generation_time_citations is not yet defined for homsap). Ideally, a specific model implementation would copy the species `generation_time_citations` into their own citations field. And analogously for the `population_size_citations`, as relevant.

It might be useful to have some explicit mechanism for marrying species attributes with a given model (specific, or not), which doesn't rely on class inheritance in the specific case.

Regarding non-specific models, `PiecewiseConstantSize` is used by the CLI when no model is specified, and the `population_size_citations` are correctly copied into the model citations in this case. Probably the `generation_time_citations` are also relevant, but are not currently copied.

**Comment by jeromekelleher** [comment]
Very good points @grahamgower. Ideally these things should be declared explicitly when defining the models, so we have attributes inside in the models that must be filled. (If not, CI fails when you try to add the model). So, a model author can either choose to use the default species generation time (and copy the citations) or they declare their own and must put in a citation for it.

It's a bit less clear in terms of the population size, since that's never really used in 'real' models, just the generic ones. So, perhaps we could manually copy the species' ``population_size_citations`` into the model's ``citations`` list? I'm not overly worried about making this work transparently in the Python API (that seems quite hard without some elaborate multiple inheritance model, which would be quite tricky and probably opaque), it's just something we need to make work for the CLI.

What do you think?

**Comment by grahamgower** [comment]
Yeah, its probably best to enforce it at model implementation time, for the specific models. The canonical way its done should also enforce coping both `param, param_citations` simultaneously (as a tuple, or its own attrib class?), where param is e.g. generation_time.

**Comment by jeromekelleher** [comment]
I think we can probably check that these are filled in as part of the automatic CI checks that should be performed when the model is added, so I don't think we need to do anything fancy. We just check to see that ``generation_time is not None`` and ``generation_time_citations is not None`` as one of the checks in the [ModelTestMixin](https://github.com/popgensims/stdpopsim/blob/master/tests/test_models.py#L15). That should do it, right?

**Comment by grahamgower** [comment]
How about something like this?
```
diff --git a/stdpopsim/catalog/homo_sapiens.py b/stdpopsim/catalog/homo_sapiens.py
index 230b958..da027b7 100644
--- a/stdpopsim/catalog/homo_sapiens.py
+++ b/stdpopsim/catalog/homo_sapiens.py
@@ -152,12 +152,6 @@ _tennessen_et_al = stdpopsim.Citation(
 # TODO: remove this superclass
 class HomoSapiensModel(stdpopsim.Model):
     species = _species
-    """
-    TODO: documentation
-    """
-    def __init__(self):
-        super().__init__()
-        self.generation_time = _species.generation_time
 
 
 # TODO we want to move away from defining these as classes which
@@ -191,7 +185,7 @@ class _GutenkunstThreePopOutOfAfrica(HomoSapiensModel):
     ]
 
     def __init__(self):
-        super().__init__()
+        super().__init__(use_species_gen_time=True)
         # First we set out the maximum likelihood values of the various parameters
         # given in Table 1.
         N_A = 7300
@@ -282,7 +276,7 @@ class _TennessenTwoPopOutOfAfrica(HomoSapiensModel):
     ]
 
     def __init__(self):
-        super().__init__()
+        super().__init__(use_species_gen_time=True)
 
         T_AF = 148e3 / self.generation_time
         T_OOA = 51e3 / self.generation_time
@@ -362,7 +356,7 @@ class _TennessenOnePopAfrica(HomoSapiensModel):
     citations = [_tennessen_et_al]
 
     def __init__(self):
-        super().__init__()
+        super().__init__(use_species_gen_time=True)
 
         T_AF = 148e3 / self.generation_time
         T_EG = 5115 / self.generation_time
@@ -426,7 +420,7 @@ class _BrowningAmerica(HomoSapiensModel):
     ]
 
     def __init__(self):
-        super().__init__()
+        super().__init__(use_species_gen_time=True)
         # Model code was ported from Supplementary File 1.
         N0 = 7310  # initial population size
         Thum = 5920  # time (gens) of advent of modern humans
diff --git a/stdpopsim/models.py b/stdpopsim/models.py
index c671684..6232909 100644
--- a/stdpopsim/models.py
+++ b/stdpopsim/models.py
@@ -162,12 +162,17 @@ class Model(object):
     # the model instances that are added to the catalog rather than
     # subclasses.
 
-    def __init__(self):
+    def __init__(self, use_species_gen_time=False):
         self.population_configurations = []
         self.demographic_events = []
         # Defaults to a single population
         self.migration_matrix = [[0]]
-        self.generation_time = -1
+
+        if use_species_gen_time:
+            if not hasattr(self, "species"):
+                raise Exception("Oooh... nasty")
+            self.generation_time = self.species.generation_time
+            self.citations.extend(self.species.generation_time_citations)
 
     @property
     def num_populations(self):
```

**Comment by jeromekelleher** [comment]
Hmmm. I'm not convinced we're getting much out of the object hierarchy here, and it'll just end up confusing people. The HomoSapiensModel is left over from a previous iteration which (IMO) was pretty convoluted.

Why not just define ``self.generation_time = None`` in the superclass, and then let implementers fill in the value when they are defining their model? The can choose to use ``_species.generation_time`` or define their own. If they forget, CI will fail.

I know there's a bit of duplication here, but the less that people coming in to add new models need to understand about the object hierarchy the better. They'll end up doing weird things anyway, so the simpler we can make it for them the easier our lives will be.

**Comment by grahamgower** [comment]
Fixed in #246
